Support JSON-like resource structures

bdenton opened the pull request:
(This is resubmission after refactoring to eliminate impact on existing interfaces)

This enhancement addresses the need to be able to serialize FHIR resources to and from streams other that Json-P or Gson. Here are two examples:

1) When building an equivalent of the jpaserver implemented in native MongoDB (no JPA or JDO), it was necessary to be able to serialize/deserialize FHIR resources via `BSONReader` and `BSONWriter`. This is possible (with some pain) with Json-P but is not at all possible with Gson.

2) We have built a general purpose data mapping engine that is able to map from a set of customer/application-specific sets of data into FHIR resources. The output of this mapping engine is a JSON-like data structure implemented in Java Maps, Lists, and scalars.

This pull request extends the `IParser` interface and updates the `JSONParser` to implement the sub-interface so that any sort of Json-like data structure or framework can be integrated in the future.

jamesagnew commented:
Hi Bill,

I'm trying to give this a review, but I'm having a lot of trouble reading it. GitHub shows that the patch touches 1761 files so it won't even put it all on the diff screen.. I'm assuming for example that you didn't actually modify android-sources.xml for example, so I think something might be wrong with the PR?

Any ideas? I'd like to get this merged but it seems like in its current form it modifies 3/4 of the repo for some reason..

bdenton commented:
James..

I am working on redoing the pull request..

What happened is that I originally submitted the PR from a branch on my fork.. that went along having build errors not related to my changes..

So, then , there ended up being a conflict in my branch so, I did an upstream merge into my local branch, fixed the conflict (not related to my change).. and re-committed and re-pushed my changed branch

This was obviously a mistake… I did not intentionally mean to change 1700+ files.. I actually only changed one file in the HAPI code stream. That one file is JsonParser. I added the package “ca.uhn.fhir.parser.json” and added the IJsonLikeParser overloaded interface… the rest is my GIT bumbling

I am kind of new to GIT and its collaboration features so any guidance on how to make changes and correctly submit pull requests would be appreciated since I obviously don’t have the correct procedure going for me.

Cheers, Bill

bdenton commented:
I am going to close this pull request yet again and try it again.. hopefully GIT will cooperate this time.
